agent/lib/kontena/observer.rb

```diff
@@ -43,33 +68,92 @@ def values
         @observables.map{|observable| @values[observable] }
       end
 
-      # Yield to block with values
+      # Still accepting updates?
+      # @return [Boolean]
+      def active?
+        fail NotImplementedError
+      end
+
+      # Run block or resume task
       def call
-        @block.call(*values)
+        fail NotImplementedError
+      end
+
+      class Async < Observe
+        # @param actor [Celluloid::Proxy::Cell] Observing actor
+        # @param block [Block] Observing block
+        def initialize(cls, observables, &block)
+          super(cls, observables)
+          @block = block
+        end
+
+        # Persistent, expected to be updated multiple times
+        # @return [Boolean]
+        def active?
+          true
+        end
+
+        def call
+          Thread.current[:celluloid_actor].task(:observe) do
+            @block.call(*values)
+          end
+        end
+      end
+
+      class Sync < Observe
+        # @param task [Celluloid::Task] Observing task
+        def initialize(cls, observables, task)
+          super(cls, observables)
+          @task = task
+          @done = false
+        end
+
+        # Single-use
+        # @return [Boolean]
+        def active?
+          !@done
+        end
+        def done!
+          @done = true
+          @task = nil
+        end
+
+        # Run block or resume task
+        def call
+          @task.resume(self)
```

Comment:
Need to deal with `@task` being nil if the `observe` timed out with a `Kontena::Observable::Message` queued... shouldn't kill the entire actor if the mailbox handler raises?